src/hotspot/share/code/codeBlob.cpp

```diff
@@ -174,8 +174,11 @@ void RuntimeBlob::free(RuntimeBlob* blob) {
 }
 
 void CodeBlob::flush() {
-  FREE_C_HEAP_ARRAY(unsigned char, _oop_maps);
-  _oop_maps = nullptr;
+  if (_oop_maps != nullptr) {
+    _oop_maps->~ImmutableOopMapSet();
+    FREE_C_HEAP_ARRAY(unsigned char, _oop_maps);
```

Comment:
Using FREE_C_HEAP_ARRAY here is concerning (even if it worked for years). CodeBlob should not know what `_oop_maps` is look like inside - he sees its type is `ImmutableOopMapSet*` and not an array.

Thank you for adding the destructor. It should free memory used by `_oop_maps`.

Reply:
If you are not comfortable to do such changes, file separate RFE to fix it later.

Reply:
>   Thank you for adding the destructor. It should free memory used by _oop_maps.

I can do that, but that would mean that the object would free its own memory it is running on which seems unusal.

Maybe it is best to add a static free_oop_map_set or similar method to ImmutableOopMapSet that frees that memory? Then that class would nominally both allocate and deallocate that object's memory.

What do you think?

Reply:
> > Thank you for adding the destructor. It should free memory used by _oop_maps.
> 
> I can do that, but that would mean that the object would free its own memory it is running on which seems unusal.
> 
> Maybe it is best to add a static free_oop_map_set or similar method to ImmutableOopMapSet that frees that memory? Then that class would nominally both allocate and deallocate that object's memory.
> 
> What do you think?

Yes, this should work too. I want ImmutableOopMapSet class handle creation and desctruction instead of Codeblob.

Reply:
How about implementing an `operator delete(void* p)` on ImmutableOopMapSet? `delete` calls an object's destructor and frees the memory at the same time. The implementation can call `FREE_C_HEAP_ARRAY`, and it avoids CodeBlob needing to call `FREE_C_HEAP_ARRAY` directly.

```
static void operator delete(void* ptr) {
  FREE_C_HEAP_ARRAY(unsigned char, ptr);
}
```

The code here would just become:
```
if (_oop_maps != nullptr) {
  delete _oop_maps;
  _oop_maps = nullptr;
}
```

Reply:
Yes, I want this!